File: semgus-sketch/src/main/java/org/semgus/sketch/frontend/Distiller.java
package org.semgus.sketch.frontend;

import org.jetbrains.annotations.NotNull;
import org.semgus.pretty.Code;
import org.semgus.sketch.base.*;

import java.util.List;
import java.util.stream.Collectors;

public interface Distiller {
  static @NotNull Code distill( Syntax x) {
    if (x == null)
      return new Code.Char('$');
    return switch (x) {
      case Id xId -> {
        var res = Code.plain(xId.name());
        for (var a : xId.idAttrs()) {
          switch (a) {
            case IdAttr.Prefix aPrefix -> res = Code.hcat(Code.plain(aPrefix.s()), res);
            case IdAttr.Postfix aPostfix -> res = Code.hcat(res, Code.plain(aPostfix.s()));
            case IdAttr.TmpStruct ignored -> res = Code.verts(res);
          }
        }
        yield res;
      }

      case Param xParam -> Code.hsep(
        distill(xParam.type()),
        distill(xParam.id())
      );

      case Expr.Ref xRef -> distill(xRef.id());
      case Expr.App xApp -> Code.hcat(
        distill(xApp.fn()),
        Code.tupled(distill(xApp.args()))
      );
      case Expr.Get xGet -> Code.hcat(
        distill(xGet.id()),
        Code.plain('.'),
        distill(xGet.field())
      );
      case Expr.Assign xAssign -> Code.hsep(
        distill(xAssign.id()),
        Code.plain('='),
        distill(xAssign.e())
      );
      case Expr.Unary xUnary -> Code.hcat(
        distill(xUnary.op()),
        Code.parens(distill(xUnary.e()))
      );
      case Expr.Binary xBinary -> Code.parens(
        Code.hsep(
          distill(xBinary.l()),
          distill(xBinary.op()),
          distill(xBinary.r())
        )
      );
      case Expr.Cond xCond -> Code.parens(
        Code.hsep(
          distill(xCond.i()),
          Code.plain('?'),
          distill(xCond.t()),
          Code.plain(':'),
          distill(xCond.e())
        )
      );
      case Expr.Choice xChoice -> Code.encloseDelim(
        Code.plain("{| "),
        Code.plain(" |}"),
        Code.plain(" | "),
        distill(xChoice.es())
      );
      // The forall expression won't be printed...
      case Expr.Forall xForall -> distill(xForall.e());

      case Stmt.VarDef xVarDef -> Code.semi(
        Code.hsep(
          distill(xVarDef.decl()),
          Code.plain('='),
          distill(xVarDef.init())
        )
      );
      case Stmt.FnDef xFnDef -> Code.block(
        Code.hcat(
          distill(xFnDef.decl()),
          Code.tupled(distill(xFnDef.params()))
        ),
        2,
        distill(xFnDef.body())
      );
      case Stmt.StructDef xStructDef -> Code.block(
        Code.hsep(
          Code.plain("struct"),
          distill(xStructDef.id())
        ),
        2,
        Code.vsep(
          distill(xStructDef.fields()).stream()
            .map(Code::semi)
            .collect(Collectors.toList())
        )
      );
      case Stmt.Atomic xAtomic -> Code.semi(
        Code.hsep(
          distill(xAtomic.prefix()),
          distill(xAtomic.e())
        )
      );
      case Stmt.Seq xSeq -> Code.vsep(
        distill(xSeq.ss())
      );
      case Stmt.Skip ignored -> Code.empty();
    };
  }

  static @NotNull List<Code> distill(@NotNull List<? extends Syntax> xs) {
    return xs.stream().map(Distiller::distill).toList();
  }
}

File: semgus-sketch/src/main/java/org/semgus/sketch/frontend/Distiller.kt
package org.semgus.sketch.frontend

import org.semgus.pretty.*
import org.semgus.sketch.syntax.*

/**
 * Frontend of sketch syntax.
 */
internal fun Syntax.distill(): Code = when (this) {
  is Id -> props.fold(plain(this.name) as Code) { acc, prop ->
    when (prop) {
      IdProp.TmpStruct -> verts(acc)
      is IdProp.Prefix -> hcat(plain(prop.s), acc)
      is IdProp.Postfix -> hcat(acc, plain(prop.s))
    }
  }

  is Param -> hsep(
    this.type.distill(),
    this.id.distill(),
  )

  is Expr.Ref -> this.id.distill()
  is Expr.Call -> hcat(
    this.fn.distill(),
    tuple0(this.args.distill()),
  )
  is Expr.Get -> hcat(
    this.obj.distill(),
    plain('.'),
    this.field.distill(),
  )
  is Expr.Assign -> hsep(
    this.l.distill(),
    plain('='),
    this.r.distill(),
  )
  is Expr.Unary -> hcat(
    plain(this.op.s),
    parens(this.e.distill()),
  )
  is Expr.Binary -> parens(
    hsep(
      this.l.distill(),
      plain(this.op.s),
      this.r.distill(),
    ),
  )
  is Expr.Nary -> hsepDelim(
    plain(this.op.s),
    this.es.distill(),
  ).let {
    if (this.es.count() == 1) it else parens(it)
  }
  is Expr.Ite -> parens(
    hsep(
      this.i.distill(),
      plain('?'),
      this.t.distill(),
      plain(':'),
      this.e.distill(),
    ),
  )
  is Expr.Choice -> hsepEncloseDelim(
    plain("{| "),
    plain(" |}"),
    plain('|'),
    this.es.distill(),
  )
  is Expr.Forall -> this.e.distill()

  Stmt.Skip -> empty()
  is Stmt.VarDef -> semi(
    hsep(
      this.decl.distill(),
      plain('='),
      this.init.distill(),
    ),
  )
  is Stmt.FnDef -> block(
    2,
    hcat(
      this.decl.distill(),
      tuple0(this.params.distill()),
    ),
    this.body.distill(),
  )
  is Stmt.StructDef -> block(
    2,
    hsep(
      plain("struct"),
      this.id.distill(),
    ),
    vsep(
      this.fields.distill().map(::semi),
    ),
  )
  is Stmt.Atomic -> semi(
    hsep(
      this.prefix.distill(),
      this.e.distill(),
    ),
  )
  is Stmt.Seq -> vsep(
    this.ss.distill(),
  )
}

internal fun Sequence<Syntax>.distill(): Sequence<Code> = this.map(Syntax::distill)
